File: Python/0206_reverse_linked_list.py
# 2020-08-05 14:44:04

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseList(self, head: ListNode) -> ListNode:
        if head is None:
            return head

        prev = None
        cur = head
        while cur is not None:
            next_ = cur.next
            cur.next = prev
            prev = cur
            cur = next_

        return prev

File: TypeScript/0206-reverse-linked-list.ts
// 2020-08-05 17:28:47

/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function reverseList(head: ListNode | null): ListNode | null {
  let prev = null;
  let cur = head;
  if (head == null) { return cur; }

  while (cur != null) { 
    let next = cur.next;
    cur.next = prev;
    prev = cur;
    cur = next
  }
  return prev;
};
